Public/Src/Utilities/KeyValueStore/RocksDb/RocksDbStore.cs

```diff
@@ -715,6 +715,73 @@ public IBuildXLKeyValueStore CreateSnapshot()
             {
                 return new RocksDbStore(this);
             }
+
+            /// <inheritdoc />
+            public IEnumerable<KeyValuePair<string, string>> PrefixSearch(string prefix, string columnFamilyName = null)
+            {
+                return PrefixSearch(StringToBytes(prefix), columnFamilyName).Select(kvp => new KeyValuePair<string, string>(BytesToString(kvp.Key), BytesToString(kvp.Value)));
+            }
+
+            /// <inheritdoc />
+            public IEnumerable<KeyValuePair<byte[], byte[]>> PrefixSearch(byte[] prefix = null, string columnFamilyName = null)
+            {
+                // TODO(jubayard): there are multiple ways to implement prefix search in RocksDB. In particular, they 
+                // have a prefix seek API (see: https://github.com/facebook/rocksdb/wiki/Prefix-Seek-API-Changes ).
+                // However, it requires certain options to be set on the column family, so it could be problematic. We
+                // just use a simpler way. Could change if any performance issues arise out of this decision.
+                var columnFamilyInfo = GetColumnFamilyInfo(columnFamilyName);
+                var readOptions = new ReadOptions();
+                readOptions.SetTotalOrderSeek(true);
+
+                using (var iterator = m_store.NewIterator(columnFamilyInfo.Handle, readOptions))
+                {
+                    if (prefix == null || prefix.Length == 0)
+                    {
+                        iterator.SeekToFirst();
+                    }
+                    else
+                    {
+                        iterator.Seek(prefix);

```

Comment:
Do we know the big-O of this Seek? Hopefully O(1) or O(ln N), not O(N).

Reply:
If the entire structure was stored in memory, then RocksDB would take O(lg N) to seek.

Given that its not, this is a very hard question to answer, especially because performance is likely dominated by disk seeks, so read amplification, cache friendliness, and others need to be taken into consideration.

In any case, the performance shouldn't be too far from a Get() call, but it is still likely to be the performance bottleneck of GetSelectors(), so if that turns into an issue we may need to spend time thinking how to optimize.

---
In reply to: [302769601](https://github.com/microsoft/BuildXL/pull/571#discussion_r302769601) [](ancestors = 302769601)